### PR TITLE
Fix: Use Var name in SubCall instead type_decl name

### LINKED ISSUE #5603: (PRIMA) use actual function instead of its type in `m_type_declaration` of Variable
It is just one step ahead of #5500 

MRE : 
```fortran
module procedure_07_module
contains
    subroutine cb(x)
        implicit none
        integer, intent(in), optional :: x(:)
    end subroutine cb

    subroutine calfun(x)
        implicit none
        integer, intent(in), optional :: x(:)
        logical :: y
        y = present(x)
        if (y) error stop
    end subroutine calfun
end module procedure_07_module

program procedure_07
    use procedure_07_module
    call temp(calfun)
contains
    subroutine temp(call_back)
        implicit none
        procedure(cb) :: call_back
        call call_back()   
    end subroutine temp
end program procedure_07
```

LFortran Error : 
```console
code generation error: Subroutine code not generated for 'cb'
Note: Please report unclear, confusing or incorrect messages as bugs at
https://github.com/lfortran/lfortran/issues.
```

### PR DESCRIPTION
**Pull request by jinangshah21**
Fixes #5603 
Refer https://github.com/lfortran/lfortran/issues/5603#issuecomment-2525200055.

### CODE REVIEW

**Review by Pranavchiku — APPROVED** [review]

File: src/libasr/codegen/asr_to_llvm.cpp [hunk 1]
```diff
@@ -9289,13 +9289,8 @@ class ASRToLLVMVisitor : public ASR::BaseVisitor<ASRToLLVMVisitor>
             proc_sym = clss_proc->m_proc;
         } else if (ASR::is_a<ASR::Variable_t>(*proc_sym)) {
             ASR::symbol_t *type_decl = ASR::down_cast<ASR::Variable_t>(proc_sym)->m_type_declaration;
-            LCOMPILERS_ASSERT(type_decl);
```
Comment (gxyd) [inline src/libasr/codegen/asr_to_llvm.cpp:9292]: these changes are straightforward and look good.

File: integration_tests/procedure_08.f90 [hunk 2]
```diff
@@ -10,6 +10,7 @@ subroutine calfun(x)
         integer, intent(in), optional :: x(:)
         logical :: y
         y = present(x)
+        print *, y
```
Comment (gxyd) [inline integration_tests/procedure_08.f90:13]: Is this acting as a test case here?

**Review by gxyd — COMMENTED** [review]
I see a change, but I don't see any case which will fail in current *main*, but pass in this PR.

**Comment by gxyd** [comment]
I'm sorry, there is a long discussion in the linked issue, but if could help me understand what exactly acts as an MRE here, that would be helpful.

**Comment by jinangshah21** [comment]
Consider MRE in https://github.com/lfortran/lfortran/issues/5603#issue-2722844558
In this when we do `call call_back()` in `temp` function it calls `cb` but in actual case it should call `calfun`. Previously call was pointing to `m_type_declaration` (which is `cb` in our MRE), but now it points to variable which will call `calfun` at runtime.

File: integration_tests/procedure_08.f90 (hunk 2 above)
Comment (jinangshah21) [reply integration_tests/procedure_08.f90:13]: It is just to make sure that `calfun` is called in our test instead of `cb`.

**Comment by gxyd** [comment]
If that's the case, do you think we should be able to add a test case signifying that it actually calls `calfun` and not `cb`?

**Comment by gxyd** [comment]
Like, maybe make `intent(inout)` in both `calfun` and `cb` and they both set the value for `x` differently, and in program use `if (...) error stop` to ensure that it's set correctly.

**Comment by certik** [comment]
Let's add an MRE that fails in main and works here.

**Review by gxyd — DISMISSED** [review]
Looks good to me.

File: integration_tests/procedure_09.f90 [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+    use procedure_09_module
+    call temp(calfun)
+contains
+    subroutine temp(call_back)
+        implicit none
+        procedure(cb) :: call_back
+        integer :: x(4)
+        call call_back(x)   
+        print *, x
+        if(x(1) /= 2) error stop
```
Comment (gxyd) [inline integration_tests/procedure_09.f90:25]: Wait a second, why can't all of this be done in `procedure_08.f90`? the file `procedure_09.f90` looks very similar.

**Review by certik — APPROVED** [review]
I think this is good, thanks!

File: integration_tests/procedure_09.f90 (hunk 3 above)
Comment (jinangshah21) [reply integration_tests/procedure_09.f90]: I have done it.

**Review by certik — APPROVED** [review]

**Comment by Pranavchiku** [comment]
Shall we merge this?

**Pull request merged**